lib/charms/kratos_external_idp_integrator/v0/kratos_external_provider.py

```diff
@@ -0,0 +1,375 @@
+}
+
+
+def get_provider_config_handler(config):
+    """Get the config handler for this provider."""
+    provider = config.get("provider")
+    if provider not in allowed_providers:
+        raise InvalidConfigError(
+            "Required configuration 'provider' MUST be one of the following: "
+            + ", ".join(allowed_providers)
+        )
+    return allowed_providers[provider]
+
+
+class RelationReadyEvent(EventBase):
+    """Event to notify the charm that the relation is ready."""
+
+    def snapshot(self):
+        """Save event."""
+        return {}
+
+    def restore(self, snapshot):
+        """Restore event."""
+        pass
+
+
+class RedirectURIChangedEvent(EventBase):
+    """Event to notify the charm that the redirect_uri changed."""
+
+    def __init__(self, handle, redirect_uri):
+        super().__init__(handle)
+        self.redirect_uri = redirect_uri
+
+    def snapshot(self):
+        """Save redirect_uri."""
+        return {"redirect_uri": self.redirect_uri}
+
+    def restore(self, snapshot):
+        """Restore redirect_uri."""
+        self.redirect_uri = snapshot["redirect_uri"]
+
+
+class ExternalIdpProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `ExternalIdpProvider`."""
+
+    ready = EventSource(RelationReadyEvent)
+    redirect_uri_changed = EventSource(RedirectURIChangedEvent)
+
+
+class ExternalIdpProvider(Object):
+    """Forward client configurations to Identity Broker."""
+
+    on = ExternalIdpProviderEvents()
+
+    def __init__(self, charm, relation_name=DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_joined, self._on_provider_endpoint_relation_joined)
+        self.framework.observe(
+            events.relation_changed, self._on_provider_endpoint_relation_changed
+        )
+        self.framework.observe(
+            events.relation_departed, self._on_provider_endpoint_relation_departed
+        )
+
+    def _on_provider_endpoint_relation_joined(self, event):
+        self.on.ready.emit()
+
+    def _on_provider_endpoint_relation_changed(self, event):
+        data = json.loads(event.relation.data[event.app]["providers"])
+        if len(data) == 0:
+            return
+        redirect_uri = data[0].get("redirect_uri")
+        self.on.redirect_uri_changed.emit(redirect_uri=redirect_uri)
+
+    def _on_provider_endpoint_relation_departed(self, event):
+        self.on.redirect_uri_changed.emit(redirect_uri="")
+
+    def is_ready(self):
+        """Checks if the relation is ready."""
+        return self._charm.model.get_relation(self._relation_name)
+
+    def create_client(self, config):
+        """Use the configuration to create the relation databag."""
+        if not self._charm.unit.is_leader():
+            return
```

Comment:
If leadership wasn't sorted out before running this method, wouldn't there be no client?

Reply:
You mean that if there is no leader unit then the relation data won't be set?

I am not sure if this is possible, according to https://juju.is/docs/sdk/a-charms-life#heading--the-graph only the `install` and `*-relation-created` are run before a leader is elected. It makes no sense for a charm to call this function in one of those two hooks. We can't have all units try to write to the databag.